### DISCUSSION TITLE
Submitting fields that are not allowed should result in a validation error

### DISCUSSION THREAD

**Discussion by MoritzLost** [craftcms/contact-form#218] [Ideas] [answered]
### Description

I like the [new `allowedMessageFields` setting](https://github.com/craftcms/contact-form/commit/20f11192b36cf3b27108d5a3a5c841d08a70659a) to limit what fields are allowed to be sent. But I've noticed that instead of adding validation so that submitting a form with illegal fields will be prevented, the mailer just [silently omits](https://github.com/craftcms/contact-form/blob/20f11192b36cf3b27108d5a3a5c841d08a70659a/src/Mailer.php#L201-L205) every field from the HTML email body that isn't allowed. So a bad actor can still tamper with the form and successfully submit it. This is troublesome for some reasons:

- If you do anything else with the submissions, like saving them somewhere, you'll still need to manually validate that there aren't any additional unallowed message fields.
- If you want to add a field to an existing form and don't know about that setting, the plugin won't provide any errors to indicate as to why the field won't show up in the e-mail. This can lead to a lot of wasted time and head-scratching.
- Manipulated requests with additional fields are likely to be malicious, so preventing those submissions completely would generally be a good idea.

Can the implementation of this feature be changed to add the `allowedMessageFields` rules as a validation rule to the `Submission` model instead?

### Steps to reproduce

1. Set the `allowedMessageFields` config to allow only some fields.
2. In any contact form, add an additional message with an input with name `message[illegalInput]`.
3. Submit the form. The submission will go through, even though the `illegalField` won't show up in the Email.

### Additional info

- Craft version: 3.7.38
- PHP version: 8.1.5
- Database driver & version: MySQL 8
- Plugins & versions: Contact Form 2.5.0

**Comment by AugustMiller** [comment, marked as answer]
We have fully reworked `message` validation for [4.x](https://github.com/craftcms/contact-form/tree/4.x); I'm not sure if we will be able to back-port this, as it could easily break existing forms that inadvertently submit additional input values.

Leaving this open, as there is likely still some hardening we can do for 3.x. :v:
